plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Escrow.hs

```diff
@@ -0,0 +1,285 @@
+escrowScript = Scripts.validatorScript . scriptInstance
+
+scriptInstance :: EscrowParams -> Scripts.ScriptInstance Escrow
+scriptInstance escrow = Scripts.Validator @Escrow
+    ($$(PlutusTx.compile [|| validate ||]) `PlutusTx.applyCode` PlutusTx.liftCode escrow)
+    $$(PlutusTx.compile [|| wrap ||])
+    where
+        wrap = Scripts.wrapValidator @PubKey @Action
+
+-- | The address of an escrow contract
+escrowAddress :: EscrowParams -> Ledger.Address
+escrowAddress = Scripts.scriptAddress . scriptInstance
+
+-- | 'pay' with an endpoint that gets the owner's public key and the
+--   contribution.
+payEp
+    ::
+    ( HasWriteTx s
+    , HasEndpoint "pay-escrow" (PubKey, Value) s
+    )
+    => EscrowParams
+    -> Contract s TxId
+payEp escrow = do
+    (ownPubKey, vl) <- endpoint @"pay-escrow"
+    pay escrow ownPubKey vl
+
+-- | Pay some money into the escrow contract.
+pay 
+    :: HasWriteTx s
+    => EscrowParams
+    -- ^ The escrow contract
+    -> PubKey
+    -- ^ Public key of the contributor (used for refunds)
+    -> Value
+    -- ^ How much money to pay in
+    -> Contract s TxId
+pay escrow ownPubKey vl = do
+    let ds = DataScript (PlutusTx.toData ownPubKey)
+        tx = payToScript vl (escrowAddress escrow) ds
+                & validityRange .~ Ledger.interval 1 (escrowDeadline escrow)
+    writeTxSuccess tx
+
+data RedeemFailReason = DeadlinePassed | NotEnoughFundsAtAddress
+  deriving (Haskell.Eq, Show)
+
+data RedeemResult =
+    RedeemSuccess TxId
+    | RedeemFail RedeemFailReason
+    deriving (Haskell.Eq, Show)
+
+-- | 'redeem' with an endpoint.
+redeemEp 
+    :: 
+    ( HasUtxoAt s 
+    , HasAwaitSlot s
+    , HasWriteTx s
+    , HasEndpoint "redeem-escrow" () s
+    )
+    => EscrowParams
+    -> Contract s RedeemResult
+redeemEp escrow = endpoint @"redeem-escrow" >> redeem escrow
+
+-- | Redeem all outputs at the contract address using a transaction that
+--   has all the outputs defined in the contract's list of targets.
+redeem 
+    :: 
+    ( HasUtxoAt s 
+    , HasAwaitSlot s
+    , HasWriteTx s
+    )
+    => EscrowParams
+    -> Contract s RedeemResult
+redeem escrow = do
+    currentSlot <- awaitSlot 0
+    unspentOutputs <- utxoAt (escrowAddress escrow)
+    let addr = escrowAddress escrow
+        valRange = 
+            let itvl = Interval.from 1
+                -- an upper bound that does not include the escrow 
+                -- deadline
+                upperBound = Interval.UpperBound (Interval.Finite $ escrowDeadline escrow) False
+            in itvl { Interval.ivTo = upperBound }
+        tx = Typed.collectFromScriptFilter (\_ _ -> True) unspentOutputs (scriptInstance escrow) Redeem
+            & validityRange .~ valRange
+            & outputs .~ fmap mkTxOutput (escrowTargets escrow)
+    if currentSlot >= escrowDeadline escrow
+    then pure $ RedeemFail DeadlinePassed
+    else if (values unspentOutputs ^. at addr . folded) `lt` targetValue escrow
```

Comment:
This needs a `gt` case too - what if an unexpected party pays into the address ("poisoning")? Then you have excess funds and the tx won't balance. I suggest:
- Send any excess funds to the person making the transaction.
- Allow the amount paid to an escrow target to be >= the specified amount so they're allowed to take the excess. The rest of the logic will make sure that everyone is paid what they're owed, so this can only allow you to take excess.